pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,262 @@
+	}
+	cacherefreshlog.Info(fmt.Sprintf("Cache initialization completed. Found %d peered clusters", len(resourceOfferList.Items)))
+	spv.peeringCache.ready = true
+	return nil
+}
+
+func (spv *Validator) refreshCache(ctx context.Context) (done bool, err error) {
+	if !spv.peeringCache.ready {
+		err = spv.initializeCache(ctx)
+		if err != nil {
+			cacherefreshlog.Info(err.Error())
+		}
+		return false, nil
+	}
+	resourceOfferList := sharing.ResourceOfferList{}
+	if err := spv.client.List(ctx, &resourceOfferList, &client.ListOptions{}); err != nil {
+		cacherefreshlog.Info(err.Error())
+		return false, nil
+	}
+	cacherefreshlog.Info("Cache refresh started")
+	// Cycling on all recorded peering in Cache
+	spv.peeringCache.peeringInfo.Range(
+		func(key, value interface{}) bool {
+			pi := value.(*peeringInfo)
+			clusterID := key.(string)
+			// Get the List of shadow pods running on the cluster
+			shadowPodList, err := spv.getShadowPodListByClusterID(ctx, clusterID)
+			if err != nil {
+				cacherefreshlog.Info(err.Error())
+				return false
+			}
+			// Flush terminating ShadowPods and check the correct alignment between cluster and cache
+			cacherefreshlog.Info(fmt.Sprintf("Aligning ShadowPods for cluster %q", pi.clusterName))
+			alignTerminatingOrNotExistingShadowPods(shadowPodList, pi)
+			return true
+		},
+	)
+
+	cacherefreshlog.Info("Aligning ResourceOffers - PeeringInfo")
+	if err := checkAlignmentResourceOfferPeeringInfo(ctx, spv); err != nil {
+		return true, err
+	}
+	cacherefreshlog.Info("Cache refresh completed")
+	return false, nil
+}
+
+func alignTerminatingOrNotExistingShadowPods(shadowPodList *vkv1alpha1.ShadowPodList, pi *peeringInfo) {
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
+	spMap := make(map[string]string)
+	cacherefreshlog.Info(fmt.Sprintf("Found %d ShadowPods for cluster %q", len(shadowPodList.Items), pi.clusterName))
+	// Check on all cluster ShadowPods and saving them in a temporary map
+	checkShadowPodsAlignment(shadowPodList, pi, spMap)
+	cacherefreshlog.Info("Searching for terminated ShadowPodDescription to be removed from cache")
+	// Alignment of all ShadowPodDescriptions in cache
+	alignTerminatingShadowPodDescriptions(pi, spMap)
+}
+
+func alignTerminatingShadowPodDescriptions(pi *peeringInfo, spMap map[string]string) {
+	for _, shadowPodDescription := range pi.spDescriptionList {
+		// Check if the ShadowPod is in terminating phase in cache and has been already terminated/deleted from the cluster
+		// if true ShadowPodDescription can be also deleted from the cache
+		if _, stillPresent := spMap[shadowPodDescription.namespacedName.String()]; !shadowPodDescription.running && !stillPresent {
+			pi.removeShadowPod(shadowPodDescription)
+			cacherefreshlog.Info(fmt.Sprintf("ShadowPodDescription %s removed from cache", shadowPodDescription.namespacedName.String()))
+		}
+	}
+}
+
+func alignExistingShadowPods(shadowPodList *vkv1alpha1.ShadowPodList, pi *peeringInfo) {
```

Comment:
Would expect this be a method of the peering info structure (same for the others).